### DISCUSSION TITLE
Scrapy + FlareSolverr

### DISCUSSION THREAD

**Discussion by the-green-man14** [FlareSolverr/FlareSolverr#876] [General]
Hello,

Has anyone successfully integrated Scrapy and FlareSolverr?

Many thanks in advance

**Comment by paulvav** [comment]
Yes, here is a simple method which you can use to get your spiders using flaresolver directly (assuming the flaresolverr container is running on your local network). Have your spider yield requests to Flaresolverr:

```py
def get_fs_body(url, method="get"):
    post_body = {
        "cmd": f"request.{method}",
        "url": url,
        "maxTimeout": 60000
    }
    return json.dumps(post_body)
```

```py
yield scrapy.Request(url=FLARE_SOLVER_URL, callback=self.callback, body=get_fs_body(url), headers={"Content-Type":"application/json"}, method="POST")
```

And then in your callback you just need to parse the flaresolverr response.
```py
def parse_fs_response(resp):
    meta = resp.meta
    resp = json.loads(resp.text)
    response = HtmlResponse(url=resp['solution']['url'], body=resp['solution']['response'], encoding='utf-8')
    return response, meta
```

```py
def callback(self, response):
    response, meta = parse_fs_response(response)
    ... rest of your callback...
```

Another option (probably better) - would be to write a download handler.

**Reply by geeone** [reply]
@the-green-man14 Sorry it took a few years :)

For anyone finding this discussion in the future: I’ve released [`scrapy-flaresolverr`](https://github.com/geeone/scrapy-flaresolverr), a Scrapy downloader middleware for integrating FlareSolverr.

It supports multiple backends, request-level options, concurrency control, and full Scrapy `HtmlResponse` reconstruction.

Hope it helps anyone looking for a maintained Scrapy + FlareSolverr integration. 😊
